balancer_test.go

```diff
@@ -408,21 +537,69 @@ func TestDefaultBalancer_Reresolve(t *testing.T) {
 	awaitPickerUpdate(t, pool, false, addrs, []int{1, 2, 3, 4})
 	conns := pool.SnapshotConns()
 	conn1 := balancertesting.FindConn(conns, addrs[0], 1)
+	require.NotNil(t, conn1)
 	conn2 := balancertesting.FindConn(conns, addrs[1], 2)
+	require.NotNil(t, conn2)
 	conn3 := balancertesting.FindConn(conns, addrs[2], 3)
+	require.NotNil(t, conn3)
 	conn4 := balancertesting.FindConn(conns, addrs[3], 4)
+	require.NotNil(t, conn4)
 	checker.UpdateHealthState(conn1, health.StateUnhealthy)
 	checker.UpdateHealthState(conn2, health.StateUnhealthy)
 	awaitResolveNow(t, pool, 1)
 	checker.UpdateHealthState(conn3, health.StateUnhealthy)
 	clock.Advance(10 * time.Second)
 	checker.UpdateHealthState(conn4, health.StateUnhealthy)
 	awaitResolveNow(t, pool, 2)
+}
 
-	err := balancer.Close()
-	require.NoError(t, err)
-	checkers := checker.SnapshotConns()
-	require.Empty(t, checkers)
+func TestBalancer_RoundTripperMaxLifetime(t *testing.T) {
+	t.Parallel()
+	clock := clocktest.NewFakeClock()
+	pool := balancertesting.NewFakeConnPool()
+
+	balancer := newBalancer(context.Background(), balancertesting.NewFakePicker, health.NopChecker, pool, time.Second)
+	balancer.updateHook = balancertesting.DeterministicReconciler
+	balancer.clock = clock
+	balancer.start()
+	t.Cleanup(func() {
+		err := balancer.Close()
+		require.NoError(t, err)
+	})
+
+	// Initial resolve
+	addrs := []resolver.Address{
+		{HostPort: "1.2.3.1"},
+		{HostPort: "1.2.3.2"},
+		{HostPort: "1.2.3.3"},
+		{HostPort: "1.2.3.4"},
+	}
+	balancer.OnResolve(addrs)
+	awaitPickerUpdate(t, pool, true, addrs, []int{1, 2, 3, 4})
+	conns := pool.SnapshotConns()
+	conn1 := balancertesting.FindConn(conns, addrs[0], 1)
+	require.NotNil(t, conn1)
+	conn2 := balancertesting.FindConn(conns, addrs[1], 2)
+	require.NotNil(t, conn2)
+	conn3 := balancertesting.FindConn(conns, addrs[2], 3)
+	require.NotNil(t, conn3)
+	conn4 := balancertesting.FindConn(conns, addrs[3], 4)
+	require.NotNil(t, conn4)
+
+	// After one second, lifetime limit is reached and all of these get recycled.
+	clock.Advance(time.Second)
+	// We can await the indices to be updated. But we won't know the precise
+	// address associated with each index because they are updated from
+	// independent timer goroutines.
+	conns = awaitConnIndices(t, pool, []int{5, 6, 7, 8})
```

Comment:
Question: do we want to "smear" the recycling over a period of time, so that we're not re-creating all connections all at the same time? It could be wise, for availability purposes, to limit the rate of recycling so that we don't inadvertently tear-down too many healthy connections that get replaced with unhealthy ones. It could also be useful for resource usage, if there are a large number of TLS connections using a cipher suite that is CPU-hungry during TLS setups.

Reply:
Hmm.. Thinking about it, it seems like it would be ideal if we could somehow rate limit the number of connections recycled within a certain period of time. That said, I am not sure what the easiest way to go about this would be that wouldn't require some ridiculous refactor.

Reply:
It might not be too bad. The backoff/rate-limiting could be in the one goroutine that processes updates (where it calls `connMgr.recycleConns`). So instead of the current approach, where each connection has its own independent timer for recycling, we could build something like Java's [`DelayQueue`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/DelayQueue.html) that is processed by that same goroutine.

But I think the more pressing question is it actually worth doing that _now_? I'm leaning towards just a TODO for now (until we or a user in the community actually encounters an issue). It's definitely not a concern for our own initial needs for this stuff. In practice, it's unlikely to become an issue unless a client is using 1000s of TLS connections (or maybe 100s of connections that use client certificates and a more computationally expensive private key algorithm like RSA).

Reply:
I guess not. Maybe would make a good TODO or issue to have in the backlog, but in practice it's probably fine, I'd suppose.